pkg/framework/unstruct/interface.go

```diff
@@ -0,0 +1,112 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package unstruct
+
+// Map lets you manipulate an object.
+type Map interface {
+	// Data returns the underlying data provided by the object.
+	Data() map[string]interface{}
+
+	// Parent returns the parent value or nil if no parent is
+	// present (root).
+	Parent() Value
+
+	// Field returns the Value object in the "key" field.
+	//
+	// If this key is not present in the map, this function will
+	// create, set and return a new Value for this field, with a nil
+	// content.
+	Field(key string) Value
+
+	// HasField returns true if the field exists.
+	HasField(key string) bool
+
+	// Fields returns the sorted list of fields in the map.
+	Fields() []string
+}
+
+// Slice lets you manipulate an array.
+type Slice interface {
+	// Data returns the underlying data provided by the object.
+	Data() []interface{}
+
+	// Parent returns the parent value or nil if no parent is
+	// present (root).
+	Parent() Value
+
+	// Length returns the number of items in the slice.
+	Length() int
+
+	// At returns the "index"-th item in the slice.
+	//
+	// If the index is out-of-bound, nil is returned.
+	At(index int) Value
+
+	// Append adds at the end of the slice if the current object is
+	// a slice.
+	//
+	// Returns a Value pointing to that item.
+	Append(value interface{}) Value
+
+	// InsertAt inserts at the index-th position in the slice.
+	//
+	// Returns the slice itself, so that the calls can be chained,
+	// or nil if the index is out-of-bound.
+	InsertAt(index int, value interface{}) Slice
```

Comment:
Yeah it would crash, but the next commit will kind of give us guarantees that some things can be done (without having one to check for nil everywhere).

Reply:
Lets stipulate that there is an implementation that will never return nil. Nevertheless, the client by definition only sees this interface.  They must eschew chaining and write nested: if x:=foo; x != nil {...} everywhere if they don't want to get a server flapping in prod because of bad input.  Given that, it makes sense to just use the Go idiom if x, ok := foo; ok {...}.  Chaining  doesn't work without an out-of-band path for errors.

Perhaps: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

Reply:
I disagree with that, but you need to see the next set of commits to get a better understanding.

Reply:
> `it makes sense to just use the Go idiom if x, ok := foo; ok {...}.` 

That and the other comment, I think we have a misunderstanding. There is a difference between "a valid `Value` points to `nil`" and "A nil `Value`".

Golang needs `if x, ok := ...` because X can't be nil (assuming it's not a pointer), and 0 is a valid value.

To phrase differently, we need three states:
- `Value` is invalid, it is `nil`,
- `Value` is valid (non-`nil`), and points to a `nil` value,
- `Value` is valid (non-`nil`), and points to a non-`nil` value,

And that's exactly what we have with the current code. No need for double return.